workers/price-reporter/src/manager/external_executor.rs

```diff
@@ -0,0 +1,467 @@
+                AtomicPriceStreamState::new_from_exchanges(&exchanges),
+            );
+        }
+
+        // Spawn a price streamer loop associated with the pair
+        self.clone().spawn_price_streamer_loop(base_token, quote_token);
+
+        Ok(())
+    }
+
+    /// Handler for the PeekPrice job
+    async fn peek_price(
+        &mut self,
+        base_token: Token,
+        quote_token: Token,
+        channel: TokioSender<PriceReporterState>,
+    ) -> Result<(), PriceReporterError> {
+        // TODO: Get-or-create price stream subscription
+        let state = self.get_state(base_token, quote_token).await;
+        channel.send(state).unwrap();
+
+        Ok(())
+    }
+
+    /// Get the state of the price reporter for the given token pair
+    async fn get_state(&self, base_token: Token, quote_token: Token) -> PriceReporterState {
+        // We don't currently support unnamed pairs
+        if !is_pair_named(&base_token, &quote_token) {
+            return PriceReporterState::UnsupportedPair(base_token, quote_token);
+        }
+
+        // Fetch the most recent Binance price
+        match self.get_latest_price(&Exchange::Binance, &base_token, &quote_token).await {
+            None => PriceReporterState::NotEnoughDataReported(0),
+            Some((price, ts)) => {
+                // Fetch the most recent prices from all other exchanges
+                let mut exchange_prices = Vec::new();
+                for exchange in ALL_EXCHANGES {
+                    if let Some((price, ts)) =
+                        self.get_latest_price(exchange, &base_token, &quote_token).await
+                    {
+                        exchange_prices.push((*exchange, (price, ts)));
+                    }
+                }
+
+                // Compute the state of the price reporter
+                compute_price_reporter_state(base_token, quote_token, price, ts, &exchange_prices)
+            },
+        }
+    }
+
+    /// Get the latest price for the given exchange and token pair
+    async fn get_latest_price(
+        &self,
+        exchange: &Exchange,
+        base_token: &Token,
+        quote_token: &Token,
+    ) -> Option<(Price, u64)> {
+        self.subscription_states
+            .read()
+            .await
+            .get(&(base_token.clone(), quote_token.clone()))
+            .and_then(|state| state.read_price(exchange))
+    }
+}
+
+/// The main loop for the websocket handler, responsible for forwarding
+/// messages between the external price reporter and the executor, and
+/// re-establishing connections indefinitely in case of failure
+async fn ws_handler_loop(
+    price_reporter_url: Url,
+    subscription_states: SubscriptionStates,
+    mut msg_out_rx: UnboundedReceiver<WebsocketMessage>,
+    msg_in_tx: UnboundedSender<PriceMessage>,
+) -> Result<(), PriceReporterError> {
+    // Outer loop handles retrying the websocket connection to the external price
+    // reporter in case of some failure
+    loop {
```

Comment:
I'd break some of this out into helpers, it's 8 nested control flow statements deep at its max

Reply:
yeah, fair criticism lol